table/block_based/block_based_table_builder.cc

```diff
@@ -1479,15 +1490,35 @@ Status BlockBasedTableBuilder::InsertBlockInCache(const Slice& block_contents,
 
     const size_t read_amp_bytes_per_bit =
         rep_->table_options.read_amp_bytes_per_bit;
-    Block* block = new Block(std::move(results), read_amp_bytes_per_bit);
-    size_t charge = block->ApproximateMemoryUsage();
-    s = block_cache->Insert(key, block, charge, &DeleteEntryCached<Block>);
-    if (s.ok()) {
-      BlockBasedTable::UpdateCacheInsertionMetrics(
-          BlockType::kData, nullptr /*get_context*/, charge,
-          s.IsOkOverwritten(), rep_->ioptions.stats);
+
+    std::unique_ptr<TBlocklike> block_holder(
+        BlocklikeTraits<TBlocklike>::Create(
+            std::move(results), read_amp_bytes_per_bit,
+            rep_->ioptions.statistics.get(),
+            false /*rep_->blocks_definitely_zstd_compressed*/,
+            rep_->table_options.filter_policy.get()));
+
+    if (block_holder->own_bytes()) {
+      size_t charge = block_holder->ApproximateMemoryUsage();
+      Cache::Handle* cache_handle = nullptr;
+
+      s = block_cache->Insert(
+          key, block_holder.get(),
+          BlocklikeTraits<TBlocklike>::GetCacheItemHelper(block_type), charge,
+          &cache_handle);
```

Comment:
We could pass ```nullptr``` here instead of ```cache_handle```, and the block cache will take care of deleting the object if it could not be inserted due to the cache being full. It'll also be slightly more efficient since we won't have to call release for every block that's inserted here.